go/src/koding/klient/machine/mount/index/index.go

```diff
@@ -0,0 +1,328 @@
+		CTime: ctimeFromSys(info),
+		MTime: info.ModTime().UnixNano(),
+		Mode:  info.Mode(),
+		Size:  info.Size(),
+		SHA1:  sum,
+	}, nil
+}
+
+// readSHA1 computes SHA-1 sum from a given file content.
+func readSHA1(path string) ([]byte, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	hash := sha1.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return nil, err
+	}
+
+	return hash.Sum(nil), nil
+}
+
+// ChangeMeta indicates what change has been done on a given file.
+type ChangeMeta uint32
+
+const (
+	ChangeMetaUpdate ChangeMeta = 1 << iota // File was updated.
+	ChangeMetaRemove                        // File was removed.
+	ChangeMetaAdd                           // File was added.
+
+	ChangeMetaLarge ChangeMeta = 1 << (8 + iota) // File size is above 4GB.
+)
+
+// Change describes single file change.
+type Change struct {
+	Name      string     `json:"name"`      // The relative name of the file.
+	Size      uint32     `json:"size"`      // Size of the file truncated to 32 bits.
+	Meta      ChangeMeta `json:"meta"`      // The type of operation made on file entry.
+	CreatedAt int64      `json:"createdAt"` // Change creation time since EPOCH.
+}
+
+// ChangeSlice stores multiple changes.
+type ChangeSlice []Change
+
+func (cs ChangeSlice) Len() int           { return len(cs) }
+func (cs ChangeSlice) Swap(i, j int)      { cs[i], cs[j] = cs[j], cs[i] }
+func (cs ChangeSlice) Less(i, j int) bool { return cs[i].Name < cs[j].Name }
+
+// Index stores a virtual working tree state. It recursively records objects in
+// a given root path and allows to efficiently detect changes on it.
+type Index struct {
+	mu      sync.RWMutex
+	entries map[string]*Entry
+}
+
+// NewIndex creates the empty index object.
+func NewIndex() *Index {
+	return &Index{
+		entries: make(map[string]*Entry, 0),
+	}
+}
+
+// NewIndexFiles walks the given file tree roted at root and records file
+// states to resulting Index object.
+func NewIndexFiles(root string) (*Index, error) {
+	idx := NewIndex()
+
+	// In order to get as much entries as we can we ignore errors.
+	walkFn := func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return nil
+		}
+
+		// Skip root path.
+		if name, err := filepath.Rel(root, path); err != nil || name == "." {
+			return nil
+		}
+
+		entry, err := NewEntryFile(root, path, info)
+		if err != nil {
+			return nil
+		}
+
+		idx.entries[entry.Name] = entry
+		return nil
+	}
+
+	if err := filepath.Walk(root, walkFn); err != nil {
+		return nil, err
+	}
+
+	return idx, nil
+}
+
+// Count returns the number of entries stored in index. Only items which size is
+// below provided value are counted. If provided argument is negative, this
+// function will return the number of all entries.
+func (idx *Index) Count(maxsize int64) (count int) {
+	idx.mu.RLock()
+	defer idx.mu.RUnlock()
+
+	if maxsize < 0 {
+		return len(idx.entries)
+	} else if maxsize == 0 {
```

Comment:
I took the idea from strings(regex) packages eg. https://golang.org/pkg/strings/#Replace where only negative numbers are treated as `all`. If this is a case I can change it.